include/SDL3/SDL_stdinc.h

```diff
@@ -80,7 +80,7 @@ char *alloca();
 /**
  *  The number of elements in an array.
  */
-#define SDL_arraysize(array)    (sizeof(array)/sizeof(array[0]))
+#define SDL_arraysize(array)    ((int)(sizeof(array)/sizeof(array[0])))
```

Comment:
If we do this, I think quite a few casts can be removed/should be added all around the code.

Reply:
yes, probably!
Also, ID should be turn to signed i think. Because they are used as to iterate arraysize
( here the list of id #7095 )

```
 190 static SDL_AudioDevice *get_audio_device(SDL_AudioDeviceID id)
 191 {
 192     id--;
 193     if ((id >= SDL_arraysize(open_devices)) || (open_devices[id] == NULL)) {
 194         SDL_SetError("Invalid audio device ID");
 195         return NULL;
 196     }
 197 
```

Reply:
I'm pretty sure amounts of memory should be a `size_t`: that's the whole point of `size_t`.

Similarly I think the result of `SDL_arraysize` makes most sense as `size_t` (which is unsigned), and therefore iterators over all items in an array also make most sense as `size_t`.

If you don't do that then you'll get into the same backwards-compatibility problems that GLib has, where the length of an array is smaller than `size_t` (in GLib it's `unsigned int`) and therefore large enough arrays will overflow the size, which is difficult to handle in a way that is neither a security vulnerability nor a crash.

Reply:
If `ssize_t` was portable, then maybe `ssize_t` would be the right type for things like lengths of arrays; but `ssize_t` isn't portable (it's `ssize_t` in POSIX, but `SSIZE_T` in Visual Studio, and I'm sure there are some crappy compilers that don't have it at all).

Using signed integers doesn't protect us from the need to detect overflow and handle it as an error: in the unsigned domain overflows wrap around (which is rarely what we would want), while in the signed domain they're undefined behaviour. So we still need things like `SDL_size_mul_overflow` or equivalent, which need implementing once per type because gcc's `__builtin_mul_overflow` isn't portable.

Reply:
The nice thing about `size_t` in this context is that by definition you can't allocate any object (array, struct, other lump of memory) larger than that, therefore *if* you know that you're counting the size of an object that already exists, it can't possibly overflow a `size_t`.

So you only need things like `SDL_size_mul_overflow` if you are asking about the size of a hypothetical object that doesn't necessarily exist yet, like a pixel buffer: if the question is "how large is my buffer?" (like for `strlen()`) then you don't need to check for overflow, but if the question is "if I had a large enough buffer to hold this width/height/bpp combination, how large would it be?" (like for `SDL_CreateRGBSurfaceWithFormat`) then you do need to check.

Reply:
I agree for all, but here, this is about static arrays pre-allocated, 
which are limited by stack size or data / bss size. I am not sure that compiler can practically let you build such array.